File: 000_C# Training/C#_11 Instances of Classes/Ruby/MainClass.rb
class MyClass

    # getters and setters are obligatory in Ruby (in VBA too, in C# not)

    @@myStaticVariable = nil

    def self.myStaticVariable
        return @@myStaticVariable
    end

    def self.myStaticVariable=value
        @@myStaticVariable = value
    end

    def myNonStaticVariable
        @myNonStaticVariable
    end

    def myNonStaticVariable=value
        @myNonStaticVariable = value
    end

    def myNonStaticVariable2
        @myNonStaticVariable2
    end

    def myNonStaticVariable2=value
        @myNonStaticVariable2 = value
    end
end

class MyClass2
    def initialize(myInt, myString)
        @myInt = myInt
        @myString = myString
    end

    def myInt
        @myInt
    end

    def myInt=value
        @myInt = value
    end

    def myString
        @myString
    end

    def myString=value
        @myString = value
    end 
end

class MyClass3
    def myInt       
        @myInt + 5
    end
    
    def myInt=value
        @myInt = value + 5
    end
end

def DeclareWriteAndRead()

    MyClass.myStaticVariable = 66

    first = MyClass.new
    second = MyClass.new

    first.myNonStaticVariable = 99
    first.myNonStaticVariable2 = 77

    second.myNonStaticVariable = 88
    second.myNonStaticVariable2 = 55        

    puts MyClass.myStaticVariable
    puts first.myNonStaticVariable
    puts first.myNonStaticVariable2
    puts second.myNonStaticVariable
    puts second.myNonStaticVariable2
end

def SaveClassInstanceForFurtherUse_List()
    # not relevant, in Ruby arrays are lists and lists and arrays
    # consult SaveClassInstanceForFurtherUse_Array()
end

def SaveClassInstanceForFurtherUse_Array()
    myList = Array.new(2)

    first = MyClass.new
    second = MyClass.new

    first.myNonStaticVariable = 99
    first.myNonStaticVariable2 = 77

    second.myNonStaticVariable = 88
    second.myNonStaticVariable2 = 55

    myList.push(first)
    myList.push(second)

    # Further use:

    myList.each do |classInstance|
        puts classInstance.myNonStaticVariable
        puts classInstance.myNonStaticVariable2
    end
end

def Initializer
    myClass2 = MyClass2.new(32, "Johann")
    puts myClass2.myInt
    puts myClass2.myString
end

def GetterSetter
    # C# does not require Getters and Setters, Java does not have them, Ruby and VBA require them
    myClass3 = MyClass3.new
    myClass3.myInt = 1
    puts myClass3.myInt
end

DeclareWriteAndRead()
SaveClassInstanceForFurtherUse_List() #obsolete
SaveClassInstanceForFurtherUse_Array()
Initializer()
GetterSetter()

File: 000_C# Training/C#_11 Instances of Classes/C#/MainClass.cs
using System;
using System.Collections.Generic;
using System.IO;

class MainClass
{
    public static void Main()
    {
        //DeclareWriteAndRead();
        //SaveClassInstanceForFurtherUse_List();
        //SaveClassInstanceForFurtherUse_Array();
        //Inherit_ExtendFunctionalityOfClass();
        Initializer();
        GetterSetter();

        Console.ReadLine();
    }

    public static void DeclareWriteAndRead()
    {
        MyClass.myStaticVariable = 66;

        MyClass first = new MyClass();
        MyClass second = new MyClass();

        first.myNonStaticVariable = 99;
        first.myNonStaticVariable2 = 77;

        second.myNonStaticVariable = 88;
        second.myNonStaticVariable2 = 55;        

        Console.WriteLine(MyClass.myStaticVariable);
        Console.WriteLine(first.myNonStaticVariable);
        Console.WriteLine(first.myNonStaticVariable2);
        Console.WriteLine(second.myNonStaticVariable);
        Console.WriteLine(second.myNonStaticVariable2);
        Console.ReadLine();    
    }

    public static void SaveClassInstanceForFurtherUse_List()
    {
        List<MyClass> myList = new List<MyClass>();

        MyClass first = new MyClass();
        MyClass second = new MyClass();

        first.myNonStaticVariable = 99;
        first.myNonStaticVariable2 = 77;

        second.myNonStaticVariable = 88;
        second.myNonStaticVariable2 = 55; 

        myList.Add(first);
        myList.Add(second);

        //Further use:

        foreach (MyClass classInstance in myList)
        {
            Console.WriteLine(classInstance.myNonStaticVariable);
            Console.WriteLine(classInstance.myNonStaticVariable2);
        }
    }

    public static void SaveClassInstanceForFurtherUse_Array()
    {
        MyClass[] myArray = new MyClass[2];

        MyClass first = new MyClass();
        MyClass second = new MyClass();

        first.myNonStaticVariable = 99;
        first.myNonStaticVariable2 = 77;

        second.myNonStaticVariable = 88;
        second.myNonStaticVariable2 = 55; 

        myArray[0] = first;
        myArray[1] = second;

        //Further use:

        for (int i=0;i<myArray.Length;i++)
        {
            Console.WriteLine(myArray[i].myNonStaticVariable);
            Console.WriteLine(myArray[i].myNonStaticVariable2);
        }
    }

    public static void Inherit_ExtendFunctionalityOfClass()
    {
        //Take all functionality of List<int> but add own (hopefully better) functionality to it.

        ListJK list = new ListJK(new int[] {6, 5, 3, 2, 5, 6, 9, 8});
        list.Add(1);
        list.Add(15);

        if (list.Contains(5))
            Console.WriteLine("Contains 5");

        list.SortJK();
        list.PrintJK();
    }

    public static void Initializer()
    {
        MyClass2 myClass2 = new MyClass2(32, "Johann");
        Console.WriteLine(myClass2.myInt);
        Console.WriteLine(myClass2.myString);
    }

    public static void GetterSetter()
    {
        // C# does not require Getters and Setters, Java does not have them, Ruby and VBA require them

        MyClass3 myClass3 = new MyClass3();
        myClass3.myInt = 1;
        Console.WriteLine(myClass3.myInt);
    }
}

class MyClass
{
    public static int myStaticVariable;
    public int myNonStaticVariable;
    public int myNonStaticVariable2;
}

class MyClass2
{
    public MyClass2(int myInt, string myString)
    {
        this.myInt = myInt;
        this.myString = myString;
    }

    public int myInt;
    public string myString;
}

class MyClass3
{
    private int _myInt;

    public int myInt
    {
        get
        {
            return _myInt + 5;
        }
        set
        {
            _myInt = value + 5;
        }
    }
}

class ListJK : List<int>
{
    public ListJK()
    {
    }

    public ListJK (int[] input)
    {
        foreach (int singleInt in input)
            base.Add(singleInt);
    }

    public void SortJK()
    {
        int maximumNumber = 0;
        int finalCounter = 0;

        for (int i=0;i<base.Count;i++)
            if (base[i] > maximumNumber)
                maximumNumber = base[i];

        int[] finalArray = new int[maximumNumber + 1];

        for (int i=0;i<base.Count;i++)
            ++finalArray[base[i]];

        for (int i=0;i<finalArray.Length;i++)
            for (int j=0;j<finalArray[i];j++)
                base[finalCounter++] = i;
    }

    public void PrintJK()
    {
        for (int i=0;i<base.Count;i++)
            Console.Write(base[i] + " ");
        
        Console.WriteLine();
    }
}
